blog/_src/posts/2016-11-02-beta-reduction-part-1.md

```diff
@@ -0,0 +1,189 @@
+
+```
+e ::= x
+   | λx.e
+   | e e
+```
+In the above BNF, x is a metavariable, standing for any variable.  In this post, I use x, y, z, a, and b as variables in my examples.  The λx.e term represents a function with a single parameter x.  We say that the parameter, x is bound in e.  It is possible to have unbound variables in the λ-calculus, though for the most part, we can ignore them in our discussion of this topic.  Finally, applications consist of two expressions.  The first expression is the function and the second is its argument.  Parenthesis are often added for clarity.
+
+If you program regularly in a functional language, this might look fairly familiar to you.  In fact, you compute anything in the λ-calculus that you can in any other language.  In other words, the λ-calculus is Turing complete.  Of course, you wouldn't want to do program in this language as it's significantly harder to encode some of these constructs than just using built in language constructs like numbers.  I'm not going to discuss these ideas in detail, but if you're interested in how to add numbers, booleans, conditionals, and recursion to the λ-calculus, Matt Might has a few great posts about how to do this using equivalent constructs in [Python](http://matt.might.net/articles/python-church-y-combinator/), [Scheme](http://matt.might.net/articles/church-encodings-demo-in-scheme/), and [JavaScript](http://matt.might.net/articles/js-church/).
+
+Now that we've discussed the syntax of the language, we can look at the semantics, or how terms evaluate.  Below I have the evaluation rules for the λ-calculus.  The arrow represents β-reduction which is the subject of this post.  The semantics rely on substitution which is depicted with brackets.  I have also defined the substitution function.  I'm assuming the Barendregt variable convention which states that every bound variable is distinct from every free variable.
+
+```
+x[ x := e ] = e
+x[ y := e ] = y
+(λx.e1)[ x := e2 ] = (λx.e1[ x := e2 ])
+(e1 e2)[ x := e3 ] = (e1[ x := e3 ] e2[ x := e3 ])
+```
+
+With the substitution function defined, we can write a semantics for evaluation:
+
+```
+------------------------------
+  (λx.e1) e2 ->β e1[ x := e2 ]
+
+    e1 ->β e1'
+--------------------
+  e1 e2 ->β e1' e2
+
+    e2 ->β e2
+--------------------
+  e1 e2 ->β e1 e2'
+```
+
+These rules mean that if you have a call in which you have a function applied to the argument, you substitute the argument into the function.  The next two rules say that if you can perform an evaluation step in either the first or second term, you can perform that step.  Notice that both the second and third rules might apply to a single term.  These rules are nondeterministic and in these cases it is fine to use whichever rule you want.
```

Comment:
That's discussed at the end of the post.  The goal of this series of posts is to discuss the proofs.  This was just introductory material that felt better as its own post.